### PR TITLE
Test Scala 3 input/output with rule built against Scala 2.12

### PR DESCRIPTION
**Pull request by bjaglin**
* Towards https://github.com/scalacenter/scalafix/issues/1316
* Requires https://github.com/scalacenter/sbt-scalafix/pull/219
* Documented in https://github.com/scalacenter/scalafix/pull/1410

This is moving the template from the built-in mutable/sequential [`crossScalaVersions`](https://www.scala-sbt.org/1.x/docs/Cross-Build.html) way to the immutable/parallel [sbt-projectmatrix](https://github.com/sbt/sbt-projectmatrix) one, to try to make it clearer that we test 3.x, but we don't cross-publish. For the record, here is the initial attempt without sbt-projectmatrix: https://github.com/scalacenter/scalafix.g8/commit/3ce2acf560a2aa2616fb67af4eaa8c667661ae05#diff-8b4a41b53e6e5f25138da6b4dc5f2d34051e59204e43fa2e9714a275997cf5e0 which [mostly](https://github.com/scalacenter/scalafix.g8/pull/19#discussion_r632880263) worked but was very hacky and hard to understand IMHO.
```
sbt:my-repo> projects
[info] In file:/home/bjaglin/git/projects/scalafix.g8/target/sbt-test/scalafix.g8-root/scripted/scalafix/
[info] 	   input2_11
[info] 	   input2_12
[info] 	   input2_13
[info] 	   input3
[info] 	   output2_11
[info] 	   output2_12
[info] 	   output2_13
[info] 	   output3
[info] 	 * my-repo
[info] 	   rules2_11
[info] 	   rules2_12
[info] 	   rules2_13
[info] 	   tests
[info] 	   testsTarget2_11
[info] 	   testsTarget2_12
[info] 	   testsTarget2_13
[info] 	   testsTarget3
```

I verified that
- the project can be imported in IntelliJ (2021.1)
- [`sbt ~tests/test`](https://github.com/scalacenter/scalafix.g8/blob/main/src/main/g8/scalafix/readme.md) is working
- `+publishLocal` publishes 3 artifacts

### CODE REVIEW

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:46]: this trick does not work if `test` is called directly under the `tests` project, so https://scalacenter.github.io/scalafix/docs/developers/setup.html should be updated if we go that way

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:51]: this trick does not work if test is called directly under the tests project, so https://scalacenter.github.io/scalafix/docs/developers/setup.html should be updated if we go that way

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:69]: https://github.com/sbt/sbt/pull/6511

File: src/main/g8/scalafix/project/plugins.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/project/plugins.sbt:2]: https://github.com/scalacenter/sbt-scalafix/pull/211

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:22]: This is required to make tests pass, but it shouldn't be. Removing it yields
```
[info]   scalafix.v1.SemanticDocument$Error$MissingSemanticdb: SemanticDB not found: input/src/main/scala/fix/MyRepo.scala
```

Despite
```
sbt:scalafix> ++ 3.0.0
[info] Setting Scala version to 3.0.0 on 3 projects.
[info] Excluded 3 projects, run ++ 3.0.0 -v for more details.
[info] Reapplying settings...
[info] set current project to scalafix (in build file:/home/bjaglin/git/projects/scalafix.g8/target/sbt-test/scalafix.g8-root/scripted/scalafix/)
sbt:scalafix> show tests/scalafixTestkitInputScalacOptions
[info] * -Xsemanticdb
[info] * -semanticdb-target
[info] * /home/bjaglin/git/projects/scalafix.g8/target/sbt-test/scalafix.g8-root/scripted/scalafix/input/target/scala-3.0.0/meta
```
```
sbt:scalafix> ++ 2.13.5
[info] Setting Scala version to 2.13.5 on 4 projects.
[info] Excluded 2 projects, run ++ 2.13.5 -v for more details.
[info] Reapplying settings...
[info] set current project to scalafix (in build file:/home/bjaglin/git/projects/scalafix.g8/target/sbt-test/scalafix.g8-root/scripted/scalafix/)
sbt:scalafix> show tests/scalafixTestkitInputScalacOptions
[info] * -P:semanticdb:synthetics:on
[info] * -Xplugin:/home/bjaglin/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scalameta/semanticdb-scalac_2.13.5/4.4.15/semanticdb-scalac_2.13.5-4.4.15.jar
[info] * -Yrangepos
[info] * -P:semanticdb:targetroot:/home/bjaglin/git/projects/scalafix.g8/target/sbt-test/scalafix.g8-root/scripted/scalafix/input/target/scala-2.13/meta

```
```
$ find . | grep MyRepo.scala.semanticdb
./input/target/scala-3.0.0/meta/META-INF/semanticdb/input/src/main/scala/fix/MyRepo.scala.semanticdb
./input/target/scala-2.12/meta/META-INF/semanticdb/input/src/main/scala/fix/MyRepo.scala.semanticdb
./input/target/scala-2.11/meta/META-INF/semanticdb/input/src/main/scala/fix/MyRepo.scala.semanticdb
./input/target/scala-2.13/meta/META-INF/semanticdb/input/src/main/scala/fix/MyRepo.scala.semanticdb

```

Comment (bjaglin) [reply src/main/g8/scalafix/build.sbt:22]: https://github.com/scalacenter/scalafix/pull/1392/files#diff-3732d353a5ef3b44d0ae648cce3bc8fdf0c3c9c666a0150fa455879f3b9e662eR34 doesn't look right, but it looks like another problem.

Comment (mlachkar) [reply src/main/g8/scalafix/build.sbt:22]: It's intersting! I struggled with semanticdb many times theses weeks (when I was developping for example the migration rule), and I had to add semanticdbInJar, I wasnt using the snapshot version, but the latest. 
I wonder if using the new api would make the detection of semanticdb files easier and clearer.

Comment (mlachkar) [reply src/main/g8/scalafix/build.sbt:22]: I will try to spend some time on this. Just to understand if it's something new or not.

Comment (bjaglin) [reply src/main/g8/scalafix/build.sbt:22]: This is only a problem with Scala 3 (the flag is parsed properly for Scala 2 before/after this PR), so the problem is either in the Scala 3 flag parsing (but the unit test looks legit), or the way `ScalafixTestKitPlugin` / `scalafix-testkit` pass the `scalaVersion` (I suspect it defaults to the runtime of the testkit, which is not 3.0.0). I will look at it again tonight as well, since I feel like it's the last part where Scala 3 support is not proper.

Comment (bjaglin) [reply src/main/g8/scalafix/build.sbt:22]: > the flag is parsed properly for Scala 2 before/after this PR

Before, it was/is tested in https://github.com/scalacenter/sbt-scalafix/commit/4278a1f4f5e7ff28d086f8431c7e64af6f5d4084 and now also in https://github.com/scalacenter/scalafix.g8/pull/19/commits/22f6e3bf345ff258a0151d73770e5f41f2fefb3e#diff-5634c415cd8c8504fdb973a3ed092300b43c4b8fc1e184f7249eb29a55511f91R13.

Comment (bjaglin) [reply src/main/g8/scalafix/build.sbt:22]: There we are: https://github.com/scalacenter/sbt-scalafix/commit/cd9efb1b1e16f178da6be72a4cc621524e8a301a#diff-9db6645e9a45c4ebf9cc4085044d892745263231d8dc14af72e1933f4a538790 - the logic is repeated there! I will check tonight (unless you beat me to it) why this is not in `scalafix-testkit` since `scalacOptions` are passed also when using the testkit...

**Comment by bjaglin** [comment]
Moving back to draft while upstreaming changed made here to https://github.com/scalacenter/sbt-scalafix/pull/219

File: src/main/g8/scalafix/project/plugins.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/project/plugins.sbt:2]: to revert when 0.9.28 is out before merging

**Comment by mlachkar** [comment]
I don't completely get why we have multiple rules modules. I think it can lead to thinking that to use write a scala 3 rule, you need to do it under rules_3 that doesn't exist.

**Comment by bjaglin** [comment]
> I don't completely get why we have multiple rules modules.

That materializes what gets cross-build, it's common when moving from `crossScalaVersions` to https://github.com/sbt/sbt-projectmatrix.

> I think it can lead to thinking that to use write a scala 3 rule, you need to do it under rules_3 that doesn't exist.

I went through several iterations to avoid that, and I thought the `testsInput3` project was quite clear. It's the central point of using https://github.com/sbt/sbt-projectmatrix: testing 4 inputs with only 3 cross-built rules. Do you have any other idea?

**Comment by mlachkar** [comment]
I would like to have @adpi2 advice/review (if you can).
Maybe we can just put rules in 2.13, no crossVersion ?

File: src/main/g8/scalafix/build.sbt
Comment (mlachkar) [inline src/main/g8/scalafix/build.sbt:35]: Is it needed ?

Comment (bjaglin) [reply src/main/g8/scalafix/build.sbt:35]: without that, commands at the root like `scalafmtAll` won't aggregate, see https://github.com/sbt/sbt-projectmatrix#two-matrices

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:73]: https://github.com/scalacenter/sbt-scalafix/pull/219/files#diff-5103eedd8d68b8d7d72de5624045280ad313a43e12e44bf6966e591963db3f62R39

Comment (bjaglin) [reply src/main/g8/scalafix/build.sbt:73]: feeding `output` to an `input`-name method is confusing, maybe the axis and the helper should be called `TestkitDataAxis` / `resolveByTestkitDataAxis`?

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:66]: added to maintain backward compatibility on CI and on [docs](https://scalacenter.github.io/scalafix/docs/developers/setup.html)

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [reply src/main/g8/scalafix/build.sbt]: I updated the name to have something more explicit than `root` (which should be a no-op compared to before)

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [reply src/main/g8/scalafix/build.sbt:73]: https://github.com/scalacenter/sbt-scalafix/pull/220

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:4]: As 3.x releases come out, we might want to test against different versions (although upgrading scala should not be as hard as on the 2.x era obviously). When/if that happens, we should use the full scala version as matrix suffixes, but I don't see a reason to do it now.

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:66]: added to maintain backward compatibility on CI statements (for people upgrading somehow) and on [docs](https://scalacenter.github.io/scalafix/docs/developers/setup.html)

**Comment by bjaglin** [comment]
CI is broken at the moment https://www.githubstatus.com/incidents/m16jzl31gnqt

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:91]: Matching the default binary version used to load scalafix in sbt-scalafix https://github.com/scalacenter/sbt-scalafix/blob/429e6181bae3c6edb4e2dd25aefa416a07f302d5/src/main/scala/scalafix/sbt/ScalafixPlugin.scala#L232

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:92]: I decided to have this by default in the matrix AND to keep the existing input/output skeletton in `scala` (instead of moving it to `scala-2`), so that people test against Scala 3 by default. That way, we can get more bug reports and fix whatever is missing in scalameta. Making the Scala3 run opt-in and/or having a `scala-2` folder in input/output would just delay the Scala3-readiness of community rules IMHO.

File: src/main/g8/scalafix/input/src/main/scala-3/fix/$repo__Camel$SignificantIndentation.scala
Comment (bjaglin) [inline src/main/g8/scalafix/input/src/main/scala-3/fix/$repo__Camel$SignificantIndentation.scala:6]: This was initially a way to test that everything works, but since we have now a scripted in `sbt-scalafix`, this is not really required. I do like the fact that it demonstrates the usage of `scala-3` intput/output.

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:35]: That's ugly, but without that `scalafmtAll` at the root wouldn't work... I am wondering if this couldn't trigger race conditions when reformatting `rules*` concurrently actually, but this is not specific to this PR, more to `sbt-projectmatrix`.

File: src/main/g8/default.properties
Comment (bjaglin) [inline src/main/g8/default.properties:1]: This way, I am hoping people name their repo `scalafix-*`, just like sbt plugins are `sbt-*`. https://github.com/liancheng/scalafix-organize-imports does, but not the others.

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:107]: about the naming https://github.com/scalacenter/sbt-scalafix/pull/220/files#r633677843

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:55]: suffix all projects for clarity, including the 2.13 one

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [inline src/main/g8/scalafix/build.sbt:88]: omit the scala version in the suffix to focus on what is being tested - the "target"

File: src/main/g8/scalafix/build.sbt
Comment (sjrd) [inline src/main/g8/scalafix/build.sbt:56]: There are two calls to `.jvmPlatform` for `input`, with different `scalaVersions`. That seems off. Shouldn't the first one already mention `rulesCrossVersions :+ scala3Version` and the second one be removed?

File: src/main/g8/scalafix/build.sbt
Comment (sjrd) [inline src/main/g8/scalafix/build.sbt:73]: What is `TestkitTargetAxis`? I can't find its definition anywhere :-s

**Review by sjrd — COMMENTED** [review]
The approach looks sound to me. I'm not sure what `TestkitTargetAxis` refers to, though :-s

**Review by mlachkar — APPROVED** [review]

**Review by sjrd — APPROVED** [review]

File: src/main/g8/scalafix/build.sbt
Comment (bjaglin) [reply src/main/g8/scalafix/build.sbt:50]: https://github.com/scalacenter/scalafix.g8/pull/19/commits/bfa2e49163dd2c715223815f832b60c5e08b8a95

**Comment by bjaglin** [comment]
waiting for 0.9.28 to appear in https://search.maven.org/solrsearch/select?q=g:%22ch.epfl.scala%22+AND+a:%22sbt-scalafix%22&rows=10&wt=xml (https://github.com/foundweekends/giter8/blob/3c90ebedbf63be384a8673c11178c53767d85617/library/src/main/scala/giter8/MavenHelper.scala#L27) for the test to pass

**Comment by bjaglin** [comment]
> waiting for 0.9.28 to appear in https://search.maven.org/solrsearch/select?q=g:%22ch.epfl.scala%22+AND+a:%22sbt-scalafix%22&rows=10&wt=xml (https://github.com/foundweekends/giter8/blob/3c90ebedbf63be384a8673c11178c53767d85617/library/src/main/scala/giter8/MavenHelper.scala#L27) for the test to pass

Still stale, so I ended up adding https://github.com/scalacenter/scalafix.g8/pull/19/commits/ee98b06db0f3726a550c90cf15152eccc3017ceb to unblock this for the release.

**Pull request merged**
